src/tokens/stream_reader.rs

```diff
@@ -0,0 +1,539 @@
+where
+    T: TokenStream<'a>,
+{
+    fn from(stream: T) -> Self {
+        TokenStreamReader {
+            stream,
+            container_stack: vec![],
+            curr_token_cell: None,
+            curr_item: StreamItem::Nothing,
+            str_buf: Str::from(""),
+        }
+    }
+}
+
+const STEP_IN_ERROR_TEXT: &str = "Cannot step in, not at start of container";
+const STEP_OUT_ERROR_TEXT: &str = "Cannot step out, not in a container";
+const NO_FIELD_NAME_ERROR_TEXT: &str = "No field name";
+const NOT_POSITIONED_ON_ANYTHING_ERROR_TEXT: &str = "Not positioned on anything";
+const CANNOT_READ_NON_SCALAR_ERROR_TEXT: &str = "Cannot read from non-scalar";
+
+/// Generates the read methods against the underlying token/item state.
+/// This eliminates the boilerplate around each definition avoiding a lot of copy/paste.
+///
+/// * `me` - the identifier for `self` to make it visible to `scalar_exp`.
+/// * `method` - the name of the method.
+/// * `scalar_type` - the return type for the method.
+/// * `variant` - the unqualified variant of [`ScalarValue`] to match over the current token.
+/// * `scalar` - the matched identifier of the contents of [`ScalarValue`].
+/// * `scalar_exp` - the expression to generate the return value.
+macro_rules! read_method_self {
+    ($me:ident, $method:ident, $scalar_type:ty, $variant:ident, $scalar:ident, $scalar_exp:expr) => {
+        fn $method(&mut $me) -> IonResult<$scalar_type> {
+            match &$me.curr_token_cell {
+                None => illegal_operation(NOT_POSITIONED_ON_ANYTHING_ERROR_TEXT),
+                Some(token_cell) => {
+                    let mut token = token_cell.borrow_mut();
+                    match token.content_mut().no_memoize_scalar() {
+                        Ok(Some(ScalarValue::$variant($scalar))) => Ok($scalar_exp),
+                        Ok(Some(scalar_value)) => illegal_operation(format!(
+                            concat!("Cannot read ", stringify!($scalar_type), " from {}"),
+                            scalar_value.scalar_type()
+                        )),
+                        Ok(None) => illegal_operation(CANNOT_READ_NON_SCALAR_ERROR_TEXT),
+                        Err(e) => Err(e),
+                    }
+                }
+            }
+        }
+    };
+}
+
+/// Shorthand for `read_method_self` where we do not need to capture `self` in the expression.
+macro_rules! read_method {
+    ($method:ident, $scalar_type:ty, $variant:ident, $scalar:ident, $scalar_exp:expr) => {
+        read_method_self!(self, $method, $scalar_type, $variant, $scalar, $scalar_exp);
+    };
+}
+
+impl<'a, T> IonReader for TokenStreamReader<'a, T>
+where
+    T: TokenStream<'a>,
+{
+    type Item = StreamItem;
+    type Symbol = Symbol;
+
+    fn ion_version(&self) -> (u8, u8) {
+        // XXX A `TokenStream` doesn't know its underlying Ion version, we just say 1.0
+        //     because all Ion 1.x versions have the same data model.
+        (1, 0)
+    }
```

Comment:
🗺️ `ElementStreamReader` has a similar weirdness--I honestly don't think this API should be here, it probably belongs at a lower-level or some other trait.